src/decoder/zlib.rs

```diff
@@ -3,10 +3,64 @@ use super::{stream::FormatErrorInner, DecodingError, CHUNCK_BUFFER_SIZE};
 use miniz_oxide::inflate::core::{decompress, inflate_flags, DecompressorOxide};
 use miniz_oxide::inflate::TINFLStatus;
 
+enum Compressor {
+    FullZlib(DecompressorOxide),
+    FDeflate(fdeflate::Decompressor),
+}
+impl Compressor {
+    fn decompress(
+        &mut self,
+        input: &[u8],
+        output: &mut [u8],
+        output_position: usize,
+        end_of_input: bool,
+    ) -> (TINFLStatus, usize, usize) {
+        match self {
+            Compressor::FullZlib(decompressor) => decompress(
+                decompressor,
+                input,
+                output,
+                output_position,
+                inflate_flags::TINFL_FLAG_PARSE_ZLIB_HEADER
+                    | inflate_flags::TINFL_FLAG_USING_NON_WRAPPING_OUTPUT_BUF
+                    | if end_of_input {
+                        0
+                    } else {
+                        inflate_flags::TINFL_FLAG_HAS_MORE_INPUT
+                    },
+            ),
+            Compressor::FDeflate(decompressor) => {
+                match decompressor.read(input, &mut output[output_position..], end_of_input) {
+                    Ok((in_consumed, out_consumed)) if decompressor.done() => {
+                        assert!(in_consumed <= input.len());
+                        (TINFLStatus::Done, in_consumed, out_consumed)
+                    }
+                    Ok((in_consumed, out_consumed)) => {
+                        assert!(in_consumed <= input.len());
+                        (TINFLStatus::HasMoreOutput, in_consumed, out_consumed)
+                    }
+                    Err(fdeflate::DecompressionError::NotFDeflate) => {
+                        *self = Compressor::FullZlib(DecompressorOxide::new());
+
+                        assert_eq!(output_position, 0);
+                        self.decompress(input, output, output_position, end_of_input)
```

Comment:
This looks wrong, the `input` is not necessarily at the start of a complete and valid `zlib` encoded stream. It is passed and conumsed in chunks. If anything, we'd need `fdeflate` to create an 'equivalent' stream header so that `DecompressorOxide` can be primed with any data that may have been contained in the already consumed header. 

That could be quite complex, as it includes:

* The zlib header
* The chunk state with header
* The last 64kB of data as a literal output, for potential backreferences
* _Optionally_ some adler32 checksum initialization, somehow

Seems complex but actually somewhat doable.

Reply:
The `fdeflate` crate is actually able to guarantee that it never consumes even a single byte of the stream before being sure that it can decode the whole thing. It does so by rejecting any deflate stream with more than one block (i.e. where BFINAL isn't set in the first block header) or whose first block describes a huffman codes it cannot support. Not sure the best way to document that guarantee.

Reply:
Huh, interesting indeed. Well, a good comment should suffice? That clears up the purpose of `assert_eq!` as well.